### PR TITLE
ci: use LLVM/Clang 17 in tidy job

### PR DESCRIPTION
**Pull request by fanquake**
Also update MSAN to use 17.0.2.

Related to #28465.

### CODE REVIEW

File: ci/test/01_base_install.sh [hunk 1]
```diff
@@ -74,8 +74,8 @@ if [[ ${USE_MEMORY_SANITIZER} == "true" ]]; then
 fi
 
 if [[ "${RUN_TIDY}" == "true" ]]; then
-  git clone --depth=1 https://github.com/include-what-you-use/include-what-you-use -b clang_16 /include-what-you-use
-  cmake -B /iwyu-build/ -G 'Unix Makefiles' -DCMAKE_PREFIX_PATH=/usr/lib/llvm-16 -S /include-what-you-use
+  git clone --depth=1 https://github.com/include-what-you-use/include-what-you-use -b master /include-what-you-use
+  cmake -B /iwyu-build/ -G 'Unix Makefiles' -DCMAKE_PREFIX_PATH=/usr/lib/llvm-17 -S /include-what-you-use
```
[the remark this answers is not in the document]
Comment (fanquake) [reply ci/test/01_base_install.sh]: Pushed up a change to do this.

**Comment by fanquake** [comment]
Given the discussion in https://github.com/include-what-you-use/include-what-you-use/issues/1276, it seems there might not actually be a new release of IWYU until a breaking change in LLVM (the master branch "seems" to be working ok here, and with LLVM 17). I wonder if we should just pick a commit from the repo, to move things forward.

**Review by maflcko — COMMENTED** [review]
Yes, seems fine to hard-code the current master commit

File: ci/test/01_base_install.sh [hunk 2]
```diff
@@ -42,7 +42,7 @@ if [ -n "$PIP_PACKAGES" ]; then
 fi
 
 if [[ ${USE_MEMORY_SANITIZER} == "true" ]]; then
-  git clone --depth=1 https://github.com/llvm/llvm-project -b "llvmorg-17.0.0-rc4" /msan/llvm-project
+  git clone --depth=1 https://github.com/llvm/llvm-project -b llvmorg-17.0.2 /msan/llvm-project
```
[the remark this answers is not in the document]
Comment (fanquake) [reply ci/test/01_base_install.sh]: Added

File: ci/test/01_base_install.sh [hunk 3]
```diff
@@ -73,8 +73,8 @@ if [[ ${USE_MEMORY_SANITIZER} == "true" ]]; then
 fi
 
 if [[ "${RUN_TIDY}" == "true" ]]; then
-  git clone --depth=1 https://github.com/include-what-you-use/include-what-you-use -b clang_16 /include-what-you-use
-  cmake -B /iwyu-build/ -G 'Unix Makefiles' -DCMAKE_PREFIX_PATH=/usr/lib/llvm-16 -S /include-what-you-use
+  git clone --depth=1 https://github.com/include-what-you-use/include-what-you-use -b master /include-what-you-use
```
[the remark this answers is not in the document]
Comment (fanquake) [reply ci/test/01_base_install.sh]: Added

**Comment by fanquake** [comment]
Tidy is green. MSAN failure is an out-of-space https://cirrus-ci.com/task/4727416686379008?logs=ci#L8613:
```bash
COMMIT ci_native_msan
Error: committing container for step {Env:[FILE_ENV=./ci/test/00_setup_env_native_msan.sh PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin] Command:run Args:[bash -c cd /ci_container_base/ && set -o errexit && source ./ci/test/00_setup_env.sh && ./ci/test/01_base_install.sh] Flags:[] Attrs:map[json:true] Message:RUN bash -c cd /ci_container_base/ && set -o errexit && source ./ci/test/00_setup_env.sh && ./ci/test/01_base_install.sh Original:RUN ["bash", "-c", "cd /ci_container_base/ && set -o errexit && source ./ci/test/00_setup_env.sh && ./ci/test/01_base_install.sh"]}: copying layers and metadata for container "9078ecd7c02b7fbc1c0de1d5d4534f78e13de447a360657750598e2941b10ce8": writing blob: adding layer with blob "sha256:4292df0af79f12a96574004b5501db4696d6f32549f7fdf6cf97f02981331925": processing tar file(write /msan/llvm-project/llvm/test/CodeGen/AMDGPU/twoaddr-regsequence.mir: no space left on device): exit status 1
```

**Pull request merged**